lib/bundle/commands/cleanup.rb

```diff
@@ -90,24 +90,31 @@ def casks_to_uninstall(global: false, file: nil)
       end
 
       def formulae_to_uninstall(global: false, file: nil)
-        @dsl ||= Brewfile.read(global:, file:)
-        kept_formulae = @dsl.entries.select { |e| e.type == :brew }.map(&:name)
-        kept_cask_formula_dependencies = Bundle::CaskDumper.formula_dependencies(kept_casks)
-        kept_formulae += kept_cask_formula_dependencies
-        kept_formulae.map! do |f|
-          Bundle::BrewDumper.formula_aliases[f] ||
-            Bundle::BrewDumper.formula_oldnames[f] ||
-            f
-        end
+        kept_formulae = self.kept_formulae(global:, file:)
 
         current_formulae = Bundle::BrewDumper.formulae
-        kept_formulae += recursive_dependencies(current_formulae, kept_formulae)
         current_formulae.reject! do |f|
           Bundle::BrewInstaller.formula_in_array?(f[:full_name], kept_formulae)
         end
         current_formulae.map { |f| f[:full_name] }
       end
 
+      def kept_formulae(global: false, file: nil)
+        return @kept_formulae if @kept_formulae
+
+        @dsl ||= Brewfile.read(global:, file:)
+
+        @kept_formulae = @dsl.entries.select { |e| e.type == :brew }.map(&:name)
+        @kept_formulae += Bundle::CaskDumper.formula_dependencies(kept_casks)
+        @kept_formulae.map! do |f|
+          Bundle::BrewDumper.formula_aliases[f] ||
+            Bundle::BrewDumper.formula_oldnames[f] ||
+            f
+        end
+
+        @kept_formulae += recursive_dependencies(Bundle::BrewDumper.formulae, kept_formulae)
```

Comment:
```suggestion
        kept_formulae = @dsl.entries.select { |e| e.type == :brew }.map(&:name)
        kept_formulae += Bundle::CaskDumper.formula_dependencies(kept_casks)
        kept_formulae.map! do |f|
          Bundle::BrewDumper.formula_aliases[f] ||
            Bundle::BrewDumper.formula_oldnames[f] ||
            f
        end

        kept_formulae += recursive_dependencies(Bundle::BrewDumper.formulae, kept_formulae)
        @kept_formulae = kept_formulae
```

to avoid setting the instance variable until we've fully populated the kept_formulae

Reply:
I changed it to use a block and also realized that I missed `@kept_formulae = nil` in the `reset!` method.